include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h

```diff
@@ -503,52 +503,78 @@ enum class __peer_prefix_algo
     scan_then_broadcast
 };
 
-template <typename _OffsetT, __peer_prefix_algo _Algo>
+template <typename _OffsetT, ::std::uint32_t __radix_bits, __peer_prefix_algo _Algo>
 struct __peer_prefix_helper;
 
-template <typename _OffsetT>
-struct __peer_prefix_helper<_OffsetT, __peer_prefix_algo::atomic_fetch_or>
+template <typename _OffsetT, ::std::uint32_t __radix_bits>
+struct __peer_prefix_helper<_OffsetT, __radix_bits, __peer_prefix_algo::atomic_fetch_or>
 {
     using _AtomicT = sycl::atomic_ref<::std::uint32_t, sycl::memory_order_relaxed, sycl::memory_scope::sub_group,
                                       sycl::access::address_space::local_space>;
     using _TempStorageT =
         sycl::accessor<::std::uint32_t, 1, sycl::access::mode::read_write, sycl::access::target::local>;
 
     sycl::sub_group __sgroup;
+    ::std::uint32_t __sg_size;
     ::std::uint32_t __self_lidx;
-    ::std::uint32_t __item_mask;
+    ::std::uint32_t __item_sg_mask;
     _AtomicT __atomic_peer_mask;
 
     __peer_prefix_helper(sycl::nd_item<1> __self_item, _TempStorageT __lacc)
-        : __sgroup(__self_item.get_sub_group()), __self_lidx(__self_item.get_local_linear_id()),
-          __item_mask(~(~0u << (__self_lidx))), __atomic_peer_mask(__lacc[0])
+        : __sgroup(__self_item.get_sub_group()),  __sg_size(__sgroup.get_local_linear_range()),
+          __self_lidx(__self_item.get_local_linear_id()), __item_sg_mask(~(~0u << (__self_lidx))),
+          __atomic_peer_mask(__lacc[0])
     {
     }
 
-    ::std::uint32_t
-    __peer_contribution(_OffsetT& __new_offset_idx, _OffsetT __offset_prefix, bool __is_current_bucket)
+    _OffsetT
+    operator()(const ::std::uint32_t __bucket_val, _OffsetT* __offset_arr)
     {
-        // reset mask for each radix state
-        if (__self_lidx == 0)
-            __atomic_peer_mask.store(0U);
+        _OffsetT __new_offset_idx = __offset_arr[__bucket_val];
         sycl::group_barrier(__sgroup);
-        // set local id's bit to 1 if the bucket value matches the radix state
-        __atomic_peer_mask.fetch_or(__is_current_bucket << __self_lidx);
+
+        // Figure out how many work-items in my subgroup have the same bucket value
+        ::std::uint32_t __peer_mask = ~((~0ull) << __sg_size);
+        for (::std::uint32_t __radix_bit = 0; __radix_bit < __radix_bits; ++__radix_bit)
+        {
+            // reset atomic for each radix bit
+            if (__self_lidx == 0)
+                __atomic_peer_mask.store(0U);
+            sycl::group_barrier(__sgroup);
```

Comment:
Just a question to clarify. In my understanding the all work items in a subgroup are executed simultaneously, 'aka' vector execution, right? if yes, why do we use an explicit barrier here?

Reply:
We use a sub-group barrier here to support devices that don't execute in lockstep.

Reply:
Yes, there are some devices where work-items in the same sub-group can be scheduled independently.

For example, threads in the same warp are not guaranteed to be synchronized at the same instruction in the Volta architecture. See https://docs.nvidia.com/cuda/volta-tuning-guide/index.html#sm-independent-thread-scheduling.